File: app/src/main/java/com/example/criminalintent/DatePickerFragment.java
package com.example.criminalintent;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.TimePicker;
import androidx.fragment.app.DialogFragment;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;

public class DatePickerFragment extends DialogFragment {

    private static final String ARG_DATE = "date";
    public static final String EXTRA_DATE = "com.bignerdranch.android.criminalintent.date";
    private DatePicker mDatePicker;
    private TimePicker mTimePicker;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Date date = (Date) getArguments().getSerializable(ARG_DATE);

        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        int year = calendar.get(Calendar.YEAR);
        int month = calendar.get(Calendar.MONTH);
        int day = calendar.get(Calendar.DAY_OF_MONTH);
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);

        View view = inflater.inflate(R.layout.dialog_date, container, false);
        mDatePicker = view.findViewById(R.id.dialog_date_picker);
        mDatePicker.init(year, month, day, null);

        mTimePicker = view.findViewById(R.id.dialog_time_picker);
        mTimePicker.setHour(hour);
        mTimePicker.setMinute(minute);

        Button okButton = view.findViewById(R.id.ok_button);
        okButton.setOnClickListener(v -> {
            int pickedYear = mDatePicker.getYear();
            int pickedMonth = mDatePicker.getMonth();
            int pickedDay = mDatePicker.getDayOfMonth();
            int pickedHour = mTimePicker.getHour();
            int pickedMinute = mTimePicker.getMinute();
            Date pickedDate = new GregorianCalendar(pickedYear, pickedMonth, pickedDay, pickedHour, pickedMinute).getTime();
            sendResult(Activity.RESULT_OK, pickedDate);
            if (getDialog() != null) {
                dismiss();
            } else {
                requireActivity().finish();
            }
        });

        return view;
    }

    public static DatePickerFragment newInstance(Date date) {
        Bundle args = new Bundle();
        args.putSerializable(ARG_DATE, date);
        DatePickerFragment fragment = new DatePickerFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private void sendResult(int resultCode, Date date) {
        Intent intent = new Intent();
        intent.putExtra(EXTRA_DATE, date);

        if (getTargetFragment() != null) {
            getTargetFragment().onActivityResult(getTargetRequestCode(), resultCode, intent);
        } else {
            requireActivity().setResult(resultCode, intent);
        }
    }
}

File: app/src/main/java/com/example/criminalintent/DatePickerFragment.kt
package com.example.criminalintent

import android.app.DatePickerDialog
import android.app.Dialog
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import java.util.Calendar
import java.util.Date

class DatePickerFragment : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val date = requireArguments().getSerializable(ARG_DATE, Date::class.java) ?: Date()
        val calendar = Calendar.getInstance().apply { time = date }

        val initialYear = calendar.get(Calendar.YEAR)
        val initialMonth = calendar.get(Calendar.MONTH)
        val initialDay = calendar.get(Calendar.DAY_OF_MONTH)

        return DatePickerDialog(
            requireContext(),
            { _, year, month, dayOfMonth ->
                val resultDate = Calendar.getInstance().apply {
                    set(Calendar.YEAR, year)
                    set(Calendar.MONTH, month)
                    set(Calendar.DAY_OF_MONTH, dayOfMonth)
                }.time

                parentFragmentManager.setFragmentResult(
                    REQUEST_KEY_DATE,
                    Bundle().apply {
                        putSerializable(BUNDLE_KEY_DATE, resultDate)
                    },
                )
            },
            initialYear,
            initialMonth,
            initialDay,
        )
    }

    companion object {
        const val REQUEST_KEY_DATE = "requestKeyDate"
        const val BUNDLE_KEY_DATE = "bundleKeyDate"

        private const val ARG_DATE = "argDate"

        fun newInstance(date: Date): DatePickerFragment {
            return DatePickerFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(ARG_DATE, date)
                }
            }
        }
    }
}
